src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectEnvVars.java

```diff
@@ -218,9 +217,16 @@ public void resolveVars(Map<String, String> variables, Map<String, String> env)
             int previousNbUnresolvedVar = nbUnresolvedVar;
             nbUnresolvedVar = 0;
             for (Map.Entry<String, String> entry : variables.entrySet()) {
-                String value = Util.replaceMacro(entry.getValue(), variables);
-                entry.setValue(value);
-                if (isUnresolvedVar(value)) {
+                String key = entry.getKey();
+                String value = entry.getValue();
+                Map<String, String> resolveVariables = new HashMap<String, String>(variables);
+                // avoid self reference expansion
+                // FOO=$FOO-X -> FOO=$FOO-X-X -> FOO=$FOO-X-X-X etc.
+                resolveVariables.remove(key);
+                String newValue = Util.replaceMacro(value, resolveVariables);
+                entry.setValue(newValue);
+                boolean variableChanged = !newValue.equals(value);
+                if (variableChanged && isUnresolvedVar(newValue)) {
```

Comment:
Are you sure that <code> variableChanged</code> is required here. For me it seems the counter won't be incremented properly if there is unresolved variables